cmd/loom/loom.go

```diff
@@ -598,14 +600,29 @@ func loadAppStore(cfg *config.Config, logger *loom.Logger, targetVersion int64)
 		if err != nil {
 			return nil, err
 		}
+		appStoreEvmRoot := iavlStore.Get(rootKey)
+
 		evmStore, err := loadEvmStore(cfg, iavlStore.Version())
 		if err != nil {
 			return nil, err
 		}
+
+		evmStoreEvmRoot, evmVersion := evmStore.GetLastSavedRoot(iavlStore.Version())
+		if !bytes.Equal(appStoreEvmRoot, evmStoreEvmRoot) && evmVersion < iavlStore.Version() {
+			// downgrade iavlStore version
+			log.Info(fmt.Sprintf("EVM roots mismatch, evm.db(%d): %X, app.db(%d): %X", evmVersion, evmStoreEvmRoot, iavlStore.Version(), appStoreEvmRoot))
+			log.Info(fmt.Sprintf("Try to load app.db at height %d", evmVersion))
+			iavlStore, err = store.NewIAVLStore(db, cfg.AppStore.MaxVersions, evmVersion, cfg.AppStore.IAVLFlushInterval)
```

Comment:
This isn't going to work reliably if `IAVLFlushInterval > 0` because the `evmVersion` may not have a corresponding IAVL tree version on disk.